In [1]:
# Some of this code was used from Harper He Code

In [1]:
# import packages
from pyspark.sql import SparkSession
from pyspark.ml import feature
from pyspark.ml import classification
from pyspark.sql import functions as fn
from pyspark.sql.functions import isnan, when, count, col
from pyspark.ml import Pipeline
from pyspark.ml.evaluation import BinaryClassificationEvaluator, \
    MulticlassClassificationEvaluator, \
    RegressionEvaluator
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
from pyspark.sql import SparkSession


spark = SparkSession.builder.getOrCreate()
sc = spark.sparkContext

In [2]:
# expand the output display to see more columns of a pandas DataFrame
pd.set_option('display.max_columns', None)  
pd.set_option('display.expand_frame_repr', False)
pd.set_option('max_colwidth', -1)

In [12]:
# This is a function for checking the shape and size
def size_shape(df):
    print("rows:",df.count())
    print("columns:",len(df.columns))

In [3]:
# Read the dataset
df = spark.read.csv('loan.csv', header=True, inferSchema=True)

In [4]:
# take a look at the first 10 rows
df_pd = df.limit(10).toPandas()
display(df_pd.head())

id member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment grade sub_grade       emp_title emp_length home_ownership annual_inc verification_status   issue_d loan_status pymnt_plan   url  desc             purpose               title zip_code addr_state    dti delinq_2yrs earliest_cr_line inq_last_6mths mths_since_last_delinq mths_since_last_record open_acc pub_rec revol_bal revol_util total_acc initial_list_status out_prncp out_prncp_inv total_pymnt total_pymnt_inv total_rec_prncp total_rec_int total_rec_late_fee recoveries collection_recovery_fee last_pymnt_d last_pymnt_amnt next_pymnt_d last_credit_pull_d collections_12_mths_ex_med mths_since_last_major_derog policy_code application_type annual_inc_joint dti_joint verification_status_joint acc_now_delinq tot_coll_amt tot_cur_bal open_acc_6m open_act_il open_il_12m open_il_24m mths_since_rcnt_il total_bal_il il_util open_rv_12m open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m  acc_open_past_24mths avg_cur_bal  bc_open_to_buy bc_util  chargeoff_within_12_mths  delinq_amnt  mo_sin_old_il_acct  mo_sin_old_rev_tl_op  mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  mort_acc  mths_since_recent_bc  mths_since_recent_bc_dlq  mths_since_recent_inq  mths_since_recent_revol_delinq  num_accts_ever_120_pd  num_actv_bc_tl  num_actv_rev_tl  num_bc_sats  num_bc_tl  num_il_tl  num_op_rev_tl  num_rev_accts  num_rev_tl_bal_gt_0  num_sats  num_tl_120dpd_2m  num_tl_30dpd  num_tl_90g_dpd_24m  num_tl_op_past_12m  pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  revol_bal_joint sec_app_earliest_cr_line  sec_app_inq_last_6mths  sec_app_mort_acc  sec_app_open_acc  sec_app_revol_util  sec_app_open_act_il  sec_app_num_rev_accts  sec_app_chargeoff_within_12_mths  sec_app_collections_12_mths_ex_med  sec_app_mths_since_last_major_derog hardship_flag hardship_type hardship_reason hardship_status deferral_term hardship_amount hardship_start_date hardship_end_date payment_plan_start_date hardship_length hardship_dpd hardship_loan_status orig_projected_additional_accrued_interest hardship_payoff_balance_amount hardship_last_payment_amount disbursement_method debt_settlement_flag debt_settlement_flag_date settlement_status settlement_date settlement_amount settlement_percentage settlement_term
0  None  None      2500       2500         2500.0            36 months  13.56     84.92        C     C1        Chef            10+ years  RENT           55000      Not Verified        Dec-2018  Current     n          None  None  debt_consolidation  Debt consolidation  109xx    NY         18.24  0           Apr-2001         1              None                   45                     9        1       4341      10.3       34        w                   2386.02   2386.02       167.02      167.02          113.98          53.04         0.0                0.0        0.0                     Feb-2019     84.92           Mar-2019     Feb-2019           0                          None                        1           Individual       None             None      None                      0              0            16901       2           2           1           2           2                  12560        69      2           7           2137       28       42000            1      11          2            9                     1878        34360.0         5.9     0.0                       0            140.0               212                   1                      1               0         1.0                  NaN                        2                     NaN                              0                      2               5                3            3          16         7              18             5                    9         0                 0             0                   3                   100.0           0.0               1                     0          60124    

In [5]:
# check how many records for each loan status
df_bystatus=df.groupby(df.loan_status).count()
df_bystatus.show()

+--------------------+-------+
|         loan_status|  count|
+--------------------+-------+
|          Fully Paid|1041952|
|             Default|     31|
|     In Grace Period|   8952|
|Does not meet the...|   1988|
|         Charged Off| 261654|
|            Oct-2015|      1|
|  Late (31-120 days)|  21897|
|             Current| 919695|
|Does not meet the...|    761|
|   Late (16-30 days)|   3737|
+--------------------+-------+



In [6]:
# show all the column names
# df.columns

In [14]:
# check the shape of df
# print("rows:",df.count())
# print("columns:",len(df.columns))

size_shape(df)

rows: 2260668
columns: 145
rows: 2260668
columns: 145


In [9]:
# check the number of Nan or Null for each column
#As per our objective, Filtering only the columns required
current_df = df.filter(col('loan_status').isin(['Late (31-120 days)','Charged Off','Late (16-30 days)','Current']))
#Renaming the Late values into a single late
current_df = current_df.withColumn("loan_status", \
              when(current_df["loan_status"].isin(['Late (31-120 days)','Late (16-30 days)']),'Late').otherwise(current_df["loan_status"]))
#Our dataset with only the desired columns values (Late,charged off,current)
print(current_df.groupby(current_df.loan_status).count().show())



+-----------+------+
|loan_status| count|
+-----------+------+
|Charged Off|261654|
|       Late| 25634|
|    Current|919695|
+-----------+------+

None


In [8]:
#Contains the nas present in all the columns in the dataset. We are planning to remove those coluomns with more than 50%Nas
checkna_pd = current_df.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in current_df.columns]).toPandas()
checkna_pd

# current_df.select([c for c in checkna_pd.columns])

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  installment  grade  sub_grade  emp_title  emp_length  home_ownership  annual_inc  verification_status  issue_d  loan_status  pymnt_plan      url     desc  purpose  title  zip_code  addr_state   dti  delinq_2yrs  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  total_acc  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  last_pymnt_d  last_pymnt_amnt  next_pymnt_d  last_credit_pull_d  collections_12_mths_ex_med  mths_since_last_major_derog  policy_code  application_type  annual_inc_joint  dti_joint  verification_status_joint  acc_now_delinq  tot_coll_amt  tot_cur_bal  open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  total_rev_hi_lim  inq_fi  total_cu_tl  inq_last_12m  acc_open_past_24mths  avg_cur_bal  bc_open_to_buy  bc_util  chargeoff_within_12_mths  delinq_amnt  mo_sin_old_il_acct  mo_sin_old_rev_tl_op  mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  mort_acc  mths_since_recent_bc  mths_since_recent_bc_dlq  mths_since_recent_inq  mths_since_recent_revol_delinq  num_accts_ever_120_pd  num_actv_bc_tl  num_actv_rev_tl  num_bc_sats  num_bc_tl  num_il_tl  num_op_rev_tl  num_rev_accts  num_rev_tl_bal_gt_0  num_sats  num_tl_120dpd_2m  num_tl_30dpd  num_tl_90g_dpd_24m  num_tl_op_past_12m  pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  revol_bal_joint  sec_app_earliest_cr_line  sec_app_inq_last_6mths  sec_app_mort_acc  sec_app_open_acc  sec_app_revol_util  sec_app_open_act_il  sec_app_num_rev_accts  sec_app_chargeoff_within_12_mths  sec_app_collections_12_mths_ex_med  sec_app_mths_since_last_major_derog  hardship_flag  hardship_type  hardship_reason  hardship_status  deferral_term  hardship_amount  hardship_start_date  hardship_end_date  payment_plan_start_date  hardship_length  hardship_dpd  hardship_loan_status  orig_projected_additional_accrued_interest  hardship_payoff_balance_amount  hardship_last_payment_amount  disbursement_method  debt_settlement_flag  debt_settlement_flag_date  settlement_status  settlement_date  settlement_amount  settlement_percentage  settlement_term
0  1206983  1206983    0          0            0                0     0         0            0      0          105247     0           0               0           0                    0        0            0           1206983  1187738  0        12520  1         1           1456  1            1                 1               622762                  1021293                 5         8        6          1120        2          3                    5          8              5            2                3                1              2                   2           2                        2414          1                261622        31                  15                          895784                       8            12                1104847           1104856    1109686                    18              10302         10299        206300       206297       206304       206307       240368              206306        355256   206310       206308       206309      206487    10316             206310  206310       206313        6884                  10370        21150           21696    39                        28           50004               10310                 10315                  10314           6885      20333                 933557                    147619                 815085                          10318                  10317           10319            8263         10317      10317      10317          10318          10317                8263      58317             10318         10316               10316  

In [11]:
#https://stackoverflow.com/questions/51322445/how-to-drop-all-columns-with-null-values-in-a-pyspark-dataframe - include in our references

#Creating a dataframe which has the count of Nas of each column
checkna = current_df.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in current_df.columns])
#Selecting the columns with NA values not greater than 50%
No_of_rows = current_df.count()
checkna_greater_than_50 = checkna.select([when(fn.col(c)<(No_of_rows*0.50),c).alias(c) for c in checkna.columns])


def drop_null_columns(df):
    """
    This function drops all columns which contain null values.
    :param df: A PySpark DataFrame
    """
    null_counts = df.select([fn.count(fn.when(fn.col(c).isNull()|isnan(fn.col(c)), c)).alias(c) for c in df.columns]).collect()[0].asDict()
    to_drop = [k for k, v in null_counts.items() if v > 0]
    df = df.drop(*to_drop)
    return df

# Using the function to find the list of columns with less than 50%Na values
final_cols = drop_null_columns(checkna_greater_than_50)
current_df1 = current_df.select([fn.col(c) for c in final_cols.columns])

#selecting our dataframe with the final cols


In [ ]:
# Benji Explore Columns 1-25

In [16]:
part1=current_df1.select(current_df1.columns[0:25])
#part1.columns
part1.limit(5).toPandas()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment grade sub_grade       emp_title emp_length home_ownership annual_inc verification_status   issue_d loan_status pymnt_plan             purpose               title zip_code addr_state    dti delinq_2yrs earliest_cr_line inq_last_6mths open_acc
0  2500       2500         2500.0            36 months  13.56     84.92        C     C1        Chef            10+ years  RENT           55000      Not Verified        Dec-2018  Current     n          debt_consolidation  Debt consolidation  109xx    NY         18.24  0           Apr-2001         1              9      
1  30000      30000        30000.0           60 months  18.94     777.23       D     D2        Postmaster      10+ years  MORTGAGE       90000      Source Verified     Dec-2018  Current     n          debt_consolidation  Debt consolidation  713xx    LA         26.52  0           Jun-1987         0              13     
2  5000       5000         5000.0            36 months  17.97     180.69       D     D1        Administrative  6 years    MORTGAGE       59280      Source Verified     Dec-2018  Current     n          debt_consolidation  Debt consolidation  490xx    MI         10.51  0           Apr-2011         0              8      
3  4000       4000         4000.0            36 months  18.94     146.51       D     D2        IT Supervisor   10+ years  MORTGAGE       92000      Source Verified     Dec-2018  Current     n          debt_consolidation  Debt consolidation  985xx    WA         16.74  0           Feb-2006         0              10     
4  30000      30000        30000.0           60 months  16.14     731.78       C     C4        Mechanic        10+ years  MORTGAGE       57250      Not Verified        Dec-2018  Current     n          debt_consolidation  Debt consolidation  212xx    MD         26.35  0           Dec-2000         0              12

In [21]:
# Dropping columns that have date or will not use in analysis
# drop the cols
drop_var=['issue_d','zip_code','earliest_cr_line']
part1=part1.drop(*drop_var)
print(size_shape(part1))
part1.limit(5).toPandas()

rows: 1206983
columns: 22
None


loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment grade sub_grade       emp_title emp_length home_ownership annual_inc verification_status loan_status pymnt_plan             purpose               title addr_state    dti delinq_2yrs inq_last_6mths open_acc
0  2500       2500         2500.0            36 months  13.56     84.92        C     C1        Chef            10+ years  RENT           55000      Not Verified        Current     n          debt_consolidation  Debt consolidation  NY         18.24  0           1              9      
1  30000      30000        30000.0           60 months  18.94     777.23       D     D2        Postmaster      10+ years  MORTGAGE       90000      Source Verified     Current     n          debt_consolidation  Debt consolidation  LA         26.52  0           0              13     
2  5000       5000         5000.0            36 months  17.97     180.69       D     D1        Administrative  6 years    MORTGAGE       59280      Source Verified     Current     n          debt_consolidation  Debt consolidation  MI         10.51  0           0              8      
3  4000       4000         4000.0            36 months  18.94     146.51       D     D2        IT Supervisor   10+ years  MORTGAGE       92000      Source Verified     Current     n          debt_consolidation  Debt consolidation  WA         16.74  0           0              10     
4  30000      30000        30000.0           60 months  16.14     731.78       C     C4        Mechanic        10+ years  MORTGAGE       57250      Not Verified        Current     n          debt_consolidation  Debt consolidation  MD         26.35  0           0              12

In [22]:
part1.printSchema()

root
 |-- loan_amnt: integer (nullable = true)
 |-- funded_amnt: integer (nullable = true)
 |-- funded_amnt_inv: double (nullable = true)
 |-- term: string (nullable = true)
 |-- int_rate: double (nullable = true)
 |-- installment: double (nullable = true)
 |-- grade: string (nullable = true)
 |-- sub_grade: string (nullable = true)
 |-- emp_title: string (nullable = true)
 |-- emp_length: string (nullable = true)
 |-- home_ownership: string (nullable = true)
 |-- annual_inc: string (nullable = true)
 |-- verification_status: string (nullable = true)
 |-- loan_status: string (nullable = true)
 |-- pymnt_plan: string (nullable = true)
 |-- purpose: string (nullable = true)
 |-- title: string (nullable = true)
 |-- addr_state: string (nullable = true)
 |-- dti: string (nullable = true)
 |-- delinq_2yrs: string (nullable = true)
 |-- inq_last_6mths: string (nullable = true)
 |-- open_acc: string (nullable = true)



In [23]:
# check the summary of each col
part1.describe().toPandas().T

0                    1                   2                                                                                                                                                                                                                                                                                                                                                            3                                                                                                                                                                                         4
summary              count    mean                 stddev              min                                                                                                                                                                                                                                                                                                                                                          max                                                                                                                                                                                     
loan_amnt            1206983  15830.35548139452    9555.658605549792   900                                                                                                                                                                                                                                                                                                                                                          40000                                                                                                                                                                                   
funded_amnt          1206983  15828.553219059424   9554.977988522249   900                                                                                                                                                                                                                                                                                                                                                          40000                                                                                                                                                                                   
funded_amnt_inv      1206983  15819.40287733609    9554.938771141058   0.0                                                                                                                                                                                                                                                                                                                                                          40000.0                                                                                                                                                                                 
term                 1206983  None                 None                 36 months                                                                                                                                                                                                                                                                                                                                                    60 months                                                                                                                                                                              
int_rate             1206983  13.464910922524924   5.05230831155585    5.31                                                                                                                                                                                                                                                                                                                                  

In [24]:
# check the missing values
checkna_part1 = part1.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in part1.columns])
checkna_part1.toPandas()

loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  installment  grade  sub_grade  emp_title  emp_length  home_ownership  annual_inc  verification_status  loan_status  pymnt_plan  purpose  title  addr_state   dti  delinq_2yrs  inq_last_6mths  open_acc
0  0          0            0                0     0         0            0      0          105247     0           0               0           0                    0            0           0        12520  1           1456  1            1               5

### Columns with missing values | the number of missing values | explanation of the type of data
emp_title      | 105,247 | The job title supplied by the Borrower when applying for the loan.*

title          | 12,520 | The loan title provided by the borrower

addr_state     | 1 | The state provided by the borrower in the loan application

dti            | 1,456 | A ratio calculated using the borrower’s total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower’s self-reported monthly income.

delinq_2yrs    | 1 | The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years

inq_last_6mths | 1 | The number of inquiries in past 6 months (excluding auto and mortgage inquiries)

open_acc       | 5 | The number of open credit lines in the borrower's credit file.


In [26]:
#